### DISCUSSION TITLE
Why does the creation of MySQL connection URLs not support the '@' special character in passwords?

### DISCUSSION THREAD

**Discussion by chungeun-choi** [sqlalchemy/sqlalchemy#10945] [Usage Questions] [answered]
When generating a URL for MySQL login with the configuration below, an incorrect URL is generated due to the regular expression in the `_parse_url` function.

```python
username = 'your_username'  
password = 'password@'  
host = 'localhost'          
port = '3306'               
database = 'your_database'
```

Logging in through the MySQL CLI does not pose any issue with this configuration. Similarly, creating a connection via SQLAlchemy should ideally not encounter problems either.

If MySQL did not support such passwords with special characters, it would be a different story. However, since MySQL does support the use of special characters in passwords, it seems necessary for SQLAlchemy to accommodate this as well.

Using the quote function appears to be a viable solution to this issue. If this hasn't been addressed yet, I am interested in contributing to the code to resolve this matter.

**Comment by CaselIT** [comment]
Hi,

You have to quote the user and password if they include special characters

**Comment by chungeun-choi** [comment]
I've read about the URL as well, but I still don't understand why it is not supported, and I'm curious to know why.
https://github.com/sqlalchemy/sqlalchemy/issues/6482

**Reply by CaselIT** [reply]
because the @ is used to indicate the start of the host

**Reply by chungeun-choi** [reply]
I'm curious why the quote function isn't used for the password when creating a URL through '_parse_url'.

**Reply by CaselIT** [reply]
I don't understand what you mean with your question.
How can `make_url` quote the password portion if you are passing it an url string? How does it know were the password starts and ends?

**Reply by Kim-Hyo-Bin** [reply]
i agree, Even though @ is frequently used in passwords, I don't know why it can't be used.
and mysql allow special character `@` 
ex) - "p@ssword"
i know using  `urllib.parse.quote_plus` but beginner coder waste the time just single character
How about adding a urllib's quote to the code before the function operates?

**Reply by CaselIT** [reply, marked as answer]
That would break the host detection. Doing `urllib.parse.quote_plus(url)` would convert `foo:bar@localhost` too making it unrecognisable.

Since you seem to have the various parts of the url as separate variables, use `URL.create` that accepts them individually, without needing to quote them

**Reply by chungeun-choi** [reply]
Sorry, looking at it again, I seem to have been mistaken. It seems difficult to handle it through a quote in the _parse_data function. The reason is that since the URL is input as a single string, it is impossible to know the delimiter once it is converted.

**Reply by chungeun-choi** [reply]
Thank you for helping me even though it was a wrong question.

**Reply by CaselIT** [reply]
no problem, sadly the fact that `@` is used both to indicate the start of the host part and that it can be in the name of the db makes it very hard to support unquoted @ in the user or password.
